src/social_integration.cpp

```diff
@@ -0,0 +1,239 @@
+			return false;
+		}
+
+		OpenTTD_SocialIntegration_v1_Init init_func = plugin->library.GetFunction("SocialIntegration_v1_Init");
+		if (plugin->library.HasError()) {
+			plugin->external.state = SocialIntegrationPlugin::UNSUPPORTED_API;
+
+			Debug(misc, 0, "[Social Integration {}] Failed to find symbol SocialPlugin_v1_Init: {}", basepath, plugin->library.GetLastError());
+			return false;
+		}
+
+		getinfo_func(&plugin->plugin_info);
+		/* Setup the information for the outside world to see. */
+		plugin->external.social_platform = plugin->plugin_info.social_platform;
+		plugin->external.name = plugin->plugin_info.name;
+		plugin->external.version = plugin->plugin_info.version;
+
+		/* Lowercase the string for comparison. */
+		std::string lc_social_platform = plugin->plugin_info.social_platform;
+		strtolower(lc_social_platform);
+
+		/* Prevent more than one plugin for a certain Social Platform to be loaded, as that never ends well. */
+		if (_loaded_social_platform.find(lc_social_platform) != _loaded_social_platform.end()) {
+			plugin->external.state = SocialIntegrationPlugin::DUPLICATE;
+
+			Debug(misc, 0, "[Social Integration: {}] Another plugin for {} is already loaded", basepath, plugin->plugin_info.social_platform);
+			return false;
+		}
+		_loaded_social_platform.insert(lc_social_platform);
+
+		auto state = init_func(&plugin->plugin_api, &plugin->openttd_info);
+		switch (state) {
+			case OTTD_SOCIAL_INTEGRATION_V1_INIT_SUCCESS:
+				plugin->external.state = SocialIntegrationPlugin::RUNNING;
+
+				Debug(misc, 1, "[Social Plugin: {}] Loaded for {}: {} ({})", basepath, plugin->plugin_info.social_platform, plugin->plugin_info.name, plugin->plugin_info.version);
+				return true;
+
+			case OTTD_SOCIAL_INTEGRATION_V1_INIT_FAILED:
+				plugin->external.state = SocialIntegrationPlugin::FAILED;
+
+				Debug(misc, 0, "[Social Plugin: {}] Failed to initialize", basepath);
+				return false;
+
+			case OTTD_SOCIAL_INTEGRATION_V1_INIT_PLATFORM_NOT_RUNNING:
+				plugin->external.state = SocialIntegrationPlugin::PLATFORM_NOT_RUNNING;
+
+				Debug(misc, 0, "[Social Plugin: {}] Failed to initialize: {} is not running", basepath, plugin->plugin_info.social_platform);
+				return false;
+
+			default:
+				NOT_REACHED();
+		}
+	}
+};
+
+std::vector<SocialIntegrationPlugin *> &SocialIntegration::GetPlugins()
+{
+	return _external_plugins;
+}
+
+void SocialIntegration::Initialize()
+{
+	SocialIntegrationFileScanner fs;
+	fs.Scan();
+}
+
+/**
+ * Wrapper to call a function pointer of a plugin if it isn't a nullptr.
+ *
+ * @param plugin Plugin to call the function pointer on.
+ * @param func   Function pointer to call.
+ */
+template <typename T, typename... Ts>
+static void PluginCall(std::unique_ptr<InternalSocialIntegrationPlugin> &plugin, T func, Ts... args)
+{
+	if (plugin->external.state != SocialIntegrationPlugin::RUNNING) {
+		return;
+	}
+
+	if (func != nullptr) {
+		func(args...);
```

Comment:
Maybe slap some try-catch block around the call, so if it's written in C++ and throws an exception and doesn't catch it by itself, then that plugin doesn't take down the whole of OpenTTD with it.
```suggestion
	try {
		func(args...);
	} catch (...) {
		Debug(misc, 0, "Social plugin {} threw an exception", plugin->plugin_info.name);
	}
```

Reply:
The API is exported as a pure C API, so we shouldn't be able to capture anything, I would think.

Reply:
I took the steam-social plugin and replaced all functions content with `throw 1`. The result, this PR crashes hard:
`terminate called after throwing an instance of 'int'`

Now with a try-catch block around the functions (I counted at least 3) you call, I get the debug message instead. So, apparently if the C code in between is thin enough it works fine for at least my current compiler. That doesn't mean it'll always work fine, but then we tried our best to keep OpenTTD running and there's nothing more to do than to tell them to remove the plugin.

Reply:
But in the same argument, if the plugin was in C, and crashes, it would still take the game down.

If (and that is a question: if) we want to solve this, I rather set a trap like we do in crashlog code, and capture any issue that might arise in the plugin handling.

But I think the question is here: do we actually want that? The plugins will be (after suggestion by 2TallTyler) guaranteed made available by us; so isn't it the same as a crash in the game, that we need to fix? Honest question, to be clear.
I just know from experience it is nearly impossible to prevent all ways a library can crash your application; so should we even try? Especially as a "throw" is rather unlikely to happen in the plugins.

Reply:
I did some further reading, and my initial instinct is correct: C++ doesn't define the ABI of anything, let alone exceptions. This is why we fall back here ofc to a C ABI, as that is well defined.

It means what you tried basically only works between a single compiler, and MIGHT work between different compilers, but also might hopefully and utterly fail in worse ways than not doing it. Even different versions of the same compiler might handle the ABI differently, making a throw from the library incompatible with the application. As it is very likely the plugins will be compiled (over time) by other compilers and versions, it is not advised to send exceptions over the barrier.

A nice write-up I found about it here:

https://wiki.sei.cmu.edu/confluence/display/cplusplus/ERR59-CPP.+Do+not+throw+an+exception+across+execution+boundaries

Here they give a nice example that OpenTTD might be build in MinGW and the plugin with MSVC. This would absolutely not work, and MSVC uses SEH for exceptions, and MinGW does not. On Linux we might be a bit more lucky, but it is just that: luck. It is UB.

That was a nice trip down a rabbit hole which I actually enjoyed :)

What is left, I think, are two options:

- Use, as with crashlog code, a `setjmp` / SEH solution to trap any issues, and recover from it.
- Or, as because of the private/public key validation I will add, consider plugins as part of our codebase that are just loaded in at a later time, and that they are no different than any other routine in our codebase. They just live in another repo.

I still vote for the last, but if there is a good argument for the first, that can be done :)